azure-kusto-data/azure/kusto/data/client.py

```diff
@@ -584,15 +599,56 @@ def __init__(self, kcsb: Union[KustoConnectionStringBuilder, str]):
 
         # Create a session object for connection pooling
         self._session = requests.Session()
-        self._session.mount("http://", HTTPAdapter(pool_maxsize=self._max_pool_size))
-        self._session.mount("https://", HTTPAdapter(pool_maxsize=self._max_pool_size))
+        adapter = HTTPAdapterWithSocketOptions(socket_options=self.compose_socket_options(), pool_maxsize=self._max_pool_size, max_retries=retries)
+        self._session.mount("http://", adapter)
+        self._session.mount("https://", adapter)
+
         self._mgmt_endpoint = "{0}/v1/rest/mgmt".format(kusto_cluster)
         self._query_endpoint = "{0}/v2/rest/query".format(kusto_cluster)
         self._streaming_ingest_endpoint = "{0}/v1/rest/ingest/".format(kusto_cluster)
         # notice that in this context, federated actually just stands for add auth, not aad federated auth (legacy code)
         self._auth_provider = _AadHelper(kcsb) if kcsb.aad_federated_security else None
         self._request_headers = {"Accept": "application/json", "Accept-Encoding": "gzip,deflate", "x-ms-client-version": "Kusto.Python.Client:" + VERSION}
 
+    @staticmethod
+    def compose_socket_options():
+        # Sends TCP Keep-Alive after MAX_IDLE_SECONDS seconds of idleness, once every INTERVAL_SECONDS seconds, and closes the connection after MAX_FAILED_KEEPALIVES failed pings (e.g. 20 => 1:00:30)
+        MAX_IDLE_SECONDS = 30
+        INTERVAL_SECONDS = 180
+        MAX_FAILED_KEEPALIVES = 20
+
+        if (
+            sys.platform == "linux"
+            and hasattr(socket, "SOL_SOCKET")
+            and hasattr(socket, "SO_KEEPALIVE")
+            and hasattr(socket, "TCP_KEEPIDLE")
+            and hasattr(socket, "TCP_KEEPINTVL")
+            and hasattr(socket, "TCP_KEEPCNT")
+        ):
+            return [

```

Comment:
This way you'll override some of the default socket options like (from urllib3): 
`#: Disable Nagle's algorithm by default.`
`#: [(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)]`
`default_socket_options = [(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)]`

My suggestion:
`HTTPConnection.default_socket_options + [`
`                                (socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1),`
`                                ...`
`                            ]`

Reply:
Good catch - I see that now in urllib3.connection.HTTPConnection. Thanks!